src/BuildIt.Forms/BuildIt.Forms.Controls/StatefulControl.xaml.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+namespace BuildIt.Forms.Controls
+{
+    [XamlCompilation(XamlCompilationOptions.Compile)]
+    public partial class StatefulControl
+    {
+        public static readonly BindableProperty StateProperty = BindableProperty.Create(nameof(State), typeof(StatefulControlStates), typeof(StatefulControl), propertyChanged: HandleStatePropertyChanged);
+
+        private const string EmptyStateContainer = nameof(EmptyStateContainer);
+        private const string ErrorStateContainer = nameof(ErrorStateContainer);
+        private const uint FadeInAnimationTimeInMilliseconds = 250;
+        private const uint FadeOutAnimationTimeInMilliseconds = FadeInAnimationTimeInMilliseconds / 2;
+        private const double FullyOpaque = 1;
+        private const double FullyTransparent = 0;
+        private const string LoadingStateContainer = nameof(LoadingStateContainer);
+
+        private readonly IDictionary<string, VisualElement> statefulContainersByName = new Dictionary<string, VisualElement>();
+        private StatefulControlStates? currentState;
+
+        public StatefulControl()
+        {
+            InitializeComponent();
+
+            ControlTemplate = new StatefulControlTemplate();
```

Comment:
Can we set this in XAML? or is there a need for it to be set in codebehind?

Reply:
Yes, of course we can. I'll move this code to XAML